packages/opentelemetry-resources/src/detectors/EnvDetector.ts

```diff
@@ -14,25 +14,142 @@
  * limitations under the License.
  */
 
-import { Detector } from '../types';
+import { Attributes, diag } from '@opentelemetry/api';
+import { getEnv } from '@opentelemetry/core';
+import { SEMRESATTRS_SERVICE_NAME } from '@opentelemetry/semantic-conventions';
 import { ResourceDetectionConfig } from '../config';
-import { IResource } from '../IResource';
-import { envDetectorSync } from './EnvDetectorSync';
+import { ResourceDetector } from '../types';
 
 /**
  * EnvDetector can be used to detect the presence of and create a Resource
  * from the OTEL_RESOURCE_ATTRIBUTES environment variable.
  */
-class EnvDetector implements Detector {
+class EnvDetector implements ResourceDetector {
+  // Type, attribute keys, and attribute values should not exceed 256 characters.
+  private readonly _MAX_LENGTH = 255;
+
+  // OTEL_RESOURCE_ATTRIBUTES is a comma-separated list of attributes.
+  private readonly _COMMA_SEPARATOR = ',';
+
+  // OTEL_RESOURCE_ATTRIBUTES contains key value pair separated by '='.
+  private readonly _LABEL_KEY_VALUE_SPLITTER = '=';
+
+  private readonly _ERROR_MESSAGE_INVALID_CHARS =
+    'should be a ASCII string with a length greater than 0 and not exceed ' +
+    this._MAX_LENGTH +
+    ' characters.';
+
+  private readonly _ERROR_MESSAGE_INVALID_VALUE =
+    'should be a ASCII string with a length not exceed ' +
+    this._MAX_LENGTH +
+    ' characters.';
+
   /**
    * Returns a {@link Resource} populated with attributes from the
    * OTEL_RESOURCE_ATTRIBUTES environment variable. Note this is an async
    * function to conform to the Detector interface.
    *
    * @param config The resource detection config
    */
-  detect(config?: ResourceDetectionConfig): Promise<IResource> {
-    return Promise.resolve(envDetectorSync.detect(config));
+  detect(_config?: ResourceDetectionConfig) {
```

Comment:
`: DetectedResource` ...if we want to be consistent about public interface return types, I'd leave a suggestion but this requires adding the import too